core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala

```diff
@@ -368,8 +301,131 @@ trait WorkspaceMigrationHistory extends DriverComponent with RawSqlQuery {
     )
 
     def withMigrationId(migrationId: Long) = selectMigrationsWhere(sql"#$idCol = $migrationId")
+  }
+
+  object migrationRetryQuery extends RawTableQuery[Long]("V1_WORKSPACE_MIGRATION_RETRIES",
+    primaryKey = ColumnName("ID")
+  ) {
+    val idCol = primaryKey
+    val migrationIdCol = ColumnName[Long]("MIGRATION_ID")
+    val retriesCol = ColumnName[Long]("RETRIES")
+
+    val allColumns: String = List(idCol, migrationIdCol, retriesCol).mkString(",")
+
+    private implicit val getWorkspaceRetry = GetResult(r => MigrationRetry(r.<<, r.<<, r.<<))
+
+    def exists(maxRetries: Int): ReadWriteAction[Boolean] =
+      nextFailedMigration(maxRetries).isDefined
+
+    final def nextFailedMigration(maxRetries: Int) = OptionT[ReadWriteAction, (Long, String, Timestamp)] {
+      sql"""
+        select m.id, w.workspaceName, m.finished from #${workspaceMigrationQuery.tableName} m
+        join (select id, CONCAT_WS("/", namespace, name) as workspaceName from WORKSPACE) as w on (w.id = m.workspace_id)
+        where m.outcome = 'Failure'
+        and (m.message like $noPermissionsFailure or m.message like $rateLimitedFailure)
+        and not exists (
+            select null from #${workspaceMigrationQuery.tableName} where workspace_id = m.workspace_id and id > m.id
+        )
+        and not exists (select null from #$tableName where #$migrationIdCol = m.id and #$retriesCol >= $maxRetries)
+        order by m.updated limit 1
+        """.as[(Long, String, Timestamp)].headOption
+    }
+
+    final def getOrCreate(migrationId: Long): ReadWriteAction[MigrationRetry] =
+      sqlu"insert ignore into #$tableName (#$migrationIdCol) values ($migrationId)" >>
+        sql"select #$allColumns from #$tableName where #$migrationIdCol = $migrationId".as[MigrationRetry].head
+
+    val noPermissionsFailure: String =
+      "%FAILED_PRECONDITION: Service account project-%@storage-transfer-service.iam.gserviceaccount.com " +
+        "does not have required permissions%"
+
+    val rateLimitedFailure: String =
+      "%RESOURCE_EXHAUSTED: Quota exceeded for quota metric 'Create requests' " +
+        "and limit 'Create requests per day' of service 'storagetransfer.googleapis.com'%"
+  }
+
+  case class RawTableQuery[PrimaryKey](tableName: String, primaryKey: ColumnName[PrimaryKey])
```

Comment:
This class is used to share the `update` functionality across the various tables